File: 1398-number-of-ways-to-stay-in-the-same-place-after-some-steps/1398-number-of-ways-to-stay-in-the-same-place-after-some-steps.java
class Solution {
    int MOD = (int)(1e9 + 7);
    Map<String, Long> dp;
    public int numWays(int steps, int arrLen) {
        dp = new HashMap<>();
        return (int)recur(steps, 0, arrLen);
    }
    public long recur(int i, int j, int k) {
        if(i == 0) {
            if(j == 0) return 1;
            return 0;
        }
        String key = i + "#" + j;
        if(dp.containsKey(key)) return dp.get(key);
        long ans = 0;
        if(j - 1 >= 0) {
            ans += recur(i - 1, j - 1, k);
            ans %= MOD;
        }
        if(j + 1 < k) {
            ans += recur(i - 1, j + 1, k);
            ans += MOD;
        }
        ans += recur(i - 1, j, k);
        ans %= MOD;
        dp.put(key, ans);
        return ans;
    }
}

File: 1398-number-of-ways-to-stay-in-the-same-place-after-some-steps/1398-number-of-ways-to-stay-in-the-same-place-after-some-steps.cpp
class Solution {
public:
    int numWays(int steps, int arrLen) {
        int MOD = 1e9 + 7;
        map<pair<int, int>, long long> mp;
        function<long long(int, int)> recur = [&](int i, int j) {
            if(i == steps) return j == 0 ? 1ll : 0ll;
            if(mp.find({i, j}) != mp.end()) return mp[{i, j}];
            long long ans = 0;
            if(j - 1 >= 0) ans += recur(i + 1, j - 1), ans %= MOD;
            if(j + 1 < arrLen) ans += recur(i + 1, j + 1), ans %= MOD;
            ans += recur(i + 1, j), ans %= MOD;
            return mp[{i, j}] = ans;
        };
        return (int)recur(0, 0);
    }
};